src/criticality/BoratedWater.C

```diff
@@ -0,0 +1,223 @@
+               "take as 5\% weight concentration or higher. Please decrease 'maximum' (" +
+                   std::to_string(_maximum) + ") to an upper limit which is in the dilute regime.");
+
+  // We take the provided material, retaining its density, but then overwriting
+  // any nuclides there to be H2O + boron weight ppm and add the S(a,b) tables
+  // for hydrogen. Therefore, we perform a check here if any nuclides in the
+  // material are *not* those we are going to add back when we wipe the material,
+  // to notify the user they probably provided the wrong material ID or they need
+  // to use a more general material modification object that allows full control
+  // over additional nuclides (e.g., if their water includes corrosion products
+  // that they do want to be there).
+
+  const int * nuclides;
+  const double * densities;
+  int n;
+  int err = openmc_material_get_densities(_material_index, &nuclides, &densities, &n);
+  catchOpenMCError(err, "get nuclide densities from material " + std::to_string(_material_id));
+
+  // get all the natural isotopes of H, O, B
+  _hydrogen_natural = NuclearData::Nuclide::getAbundances("H");
+  _boron_natural = NuclearData::Nuclide::getAbundances("B");
+  _oxygen_natural = NuclearData::Nuclide::getAbundances("O");
+
+  std::vector<std::string> allowable;
+  for (const auto & i : _hydrogen_natural)
+    allowable.push_back(i.first);
+  for (const auto & i : _boron_natural)
+    allowable.push_back(i.first);
+  for (const auto & i : _oxygen_natural)
+    allowable.push_back(i.first);
+
+  if (isParamValid("absent_nuclides"))
+  {
+    const auto & absent = getParam<std::vector<std::string>>("absent_nuclides");
+
+    for (const auto & a : absent)
+    {
+      // only missing nuclides for H, O, B are meaningful
+      if (std::find(allowable.begin(), allowable.end(), a) == allowable.end())
+        paramWarning("absent_nuclides",
+                     "Only absent isotopes of hydrogen, oxygen, or boron will be used to adjust "
+                     "natural abundances. The entry '" +
+                         a + "' will be unused.");
+
+      // adjust the natural abundances if nuclides are missing from the cross section library
+      // TODO: implement in a general fashion, this only works for O18 which is also the only
+      // absent nuclide we expect in practice
+      if (a == "O18")
+      {
+        // find which entry in _oxygen_natural has O16 and O18; this allows these isotopes
+        // to appear in any order
+        unsigned int idx16;
+        unsigned int idx18;
+        unsigned int idx = 0;
+        for (const auto & o : _oxygen_natural)
+        {
+          if (o.first == "O16")
+            idx16 = idx;
+          idx++;
+        }
+
+        idx = 0;
+        for (const auto & o : _oxygen_natural)
+        {
+          if (o.first == "O18")
+            idx18 = idx;
+          idx++;
+        }
```

Comment:
Can't these be combined?
```suggestion
        for (const auto & o : _oxygen_natural)
        {
          if (o.first == "O16")
            idx16 = idx;
          if (o.first == "O18")
            idx18 = idx;
          idx++;
        }
```